Source/WebCore/platform/mediastream/webrtcorg/PeerConnectionBackendWebRtcOrg.cpp

```diff
@@ -43,71 +61,169 @@ void enableWebRtcOrgPeerConnectionBackend()
     ASSERT(!m_sessionDescriptionPromise);
+    int requestId = generateNextId();
+    
+    webrtc::PeerConnectionInterface::RTCOfferAnswerOptions wrtcOptions;
+    wrtcOptions.voice_activity_detection = options.voiceActivityDetection();
+    webrtc::CreateSessionDescriptionObserver* observer =
+            new rtc::RefCountedObject<WRTCCreateSessionDescriptionObserver>(this, requestId);
 
-    WRTCInt::RTCOfferAnswerOptions rtcOptions;
-    rtcOptions[WRTCInt::kVoiceActivityDetection] = options.voiceActivityDetection();
-
-    int id = m_rtcConnection->createAnswer(rtcOptions);
-    if (WRTCInt::InvalidRequestId != id) {
-        m_sessionDescriptionRequestId = id;
-        m_sessionDescriptionPromise = WTFMove(promise);
-    } else {
-        promise.reject(DOMError::create("Failed to create answer"));
-    }
+    m_peerConnection->CreateAnswer(observer, wrtcOptions);
+    m_sessionDescriptionRequestId = requestId;
+    m_sessionDescriptionPromise = WTFMove(promise);
 }
 
 void PeerConnectionBackendWebRtcOrg::setLocalDescription(RTCSessionDescription& desc, PeerConnection::VoidPromise&& promise)
 {
-    ASSERT(WRTCInt::InvalidRequestId == m_voidRequestId);
+    UNUSED_PARAM(promise);
+    ASSERT(InvalidRequestId == m_voidRequestId);
     ASSERT(!m_voidPromise);
-
-    WRTCInt::RTCSessionDescription localDesc;
-    localDesc.type = desc.type().utf8().data();
-    localDesc.sdp = desc.sdp().utf8().data();
-
-    int id = m_rtcConnection->setLocalDescription(localDesc);
-    if (WRTCInt::InvalidRequestId != id) {
-        m_voidRequestId = id;
-        m_voidPromise = WTFMove(promise);
-    } else {
-        promise.reject(DOMError::create("Failed to parse local description"));
+    int requestId = generateNextId();
+    webrtc::SdpParseError error;
+    webrtc::SetSessionDescriptionObserver* observer =
+            new rtc::RefCountedObject<WRTCSetSessionDescriptionObserver>(this, requestId);
+
+    webrtc::SessionDescriptionInterface* wrtcDesc = webrtc::CreateSessionDescription(
+            desc.type().utf8().data(),
+            desc.sdp().utf8().data(),
+            &error);
+    if (!wrtcDesc) {
+        LOG(Media, "Failed to create session description, error=%s line=%s", error.description.c_str(),error.line.c_str());
+	m_voidRequestId = InvalidRequestId;
+	return;
     }
+
+    m_peerConnection->SetLocalDescription(observer, wrtcDesc);
+    m_voidRequestId = requestId;
+    m_voidPromise = WTFMove(promise);
 }
 RefPtr<RTCSessionDescription> PeerConnectionBackendWebRtcOrg::localDescription() const
 {
     // TODO: pendingLocalDescription/currentLocalDescription
-    WRTCInt::RTCSessionDescription localDesc;
-    m_rtcConnection->localDescription(localDesc);
-    String type = localDesc.type.c_str();
-    String sdp = localDesc.sdp.c_str();
-    return RTCSessionDescription::create(type, sdp);
+    const webrtc::SessionDescriptionInterface* wrtcDesc = m_peerConnection->local_description();
+    std::string sdp;
+    wrtcDesc->ToString(&sdp);
```

Comment:
In case of failure, should the function return nullptr, or take any other action?